### ISSUE TITLE
All dependencies are preloaded no matter if they are used or not

### ISSUE THREAD

**Issue by voroniys** [inspec/inspec#3095]
### 🎛 Description

bug report

It looks like inspec always preload all profiles that are mentioned in depends: section of inspec.yml
If then in the logic some profiles are excluded they are still reported with no test executed. But what is worse - the are preloading resources, for instance parsing files and so on. If some of these files are big and complex it introduces quite a performance penalty depennd on complexity of profiles.
Fore example:
in inspec.yaml I have:
```yaml
depends:
 - name: Profile_A
   path: ../A
 - name: Profile_B
   path: ../B
 - name: Profile_C
   path: ../C
 - name: Profile_D
   path: ../D
```
than in controls/my_profile.rb:
```ruby
type=<some expression>
case type
when 'A'
  include_controls 'A'
when 'B'
  include_controls 'B'
when 'C'
  include_controls 'C' 
else
  include_controls 'D'
end
```
As you can see I always need only one profile, while all 4 are always preloaded and pre-processed.
As current Inspec in fact does not support properly cascaded includes the construction mentioned above is the only way how to workaround this shortcoming and  keep inclusion depth limited to one level

### 🌍 InSpec and Platform Version

inspec --version
2.1.72

### 🤔 Replication Case

See above

### 💁 Possible Solutions

Information in inspec.yml should only be used to know the location of the profile, that may be needed, but not pre-load them. Profile has to be loaded when first include_controls/require_controls does mention it.
As current Inspec in fact does not support properly cascaded includes the construction mentioned above is the only way how to workaround this shortcoming and  keep inclusion depth limited to one level

**Comment by frezbo** [comment]
I think that's the expected beahviour. If you have a test that has something like this:
```
data = command('ls').stdout
```

it will preload as inspec has also a compile/run phase AFAIK. Maybe the maintainers can give a clear idea.

**Comment by voroniys** [comment]
Yes, I know that it is how current inspec is implemented. But my point is that implementation is fundamentally wrong from point of view of performance. Especially wrong because source of the included profile can be URL, profile can be quite big and outside connection can be poor. A lot of effort to download profile and preload them will be done in vain.

My point is that 'load' in runner should load only current profile itself and not all profile mentioned as dependencies. Dependencies should be loaded in execution phase when include_controls/require_controls is found. It should be equal to how chef-client itself is working - load current profile, than "compilation" phase that builds a collection of tests, and than execution of the collection.

IMHO current inspec implementation is exactly missing compilation phase and have a blind load phase instead.

**Comment by frezbo** [comment]
you mean it should only download if explicitly loaded as `include_controls`.?

**Comment by voroniys** [comment]
yes, exactly!
But they are not only downloaded, the are preprocessed. That is how I fond this bug - I was getting a warning message (see #3090) which never should appear, because control that cheks krb5.conf file was in another case alternative and ahould never be executed. So this means that ini() resource was executed and file was loaded and parsed for nothing!

**Comment by aaronlippold** [comment]
we have run into this as well. Right now it is an artifact of how inheritance works with InSpec. The InSpec team is working on a rebuild of inhertiance of InSpec in general and this issue is on the roadmap for resolution. @jquick @arlimus @chris-rock can we make sure this is added to the RFC.

**Comment by clintoncwolfe** [comment]
One partial workaround is to use `inspec vendor`; you mention that the download phase is painful.  `vendor` will download the profiles ahead of time, and then load them locally on the next run.  It will still execute/parse unused profiles, but it will skip the network effort.

**Comment by voroniys** [comment]
The network issues can be just additional factor. In my situation all profiles are local, but from ~1m15s total execution time approx. 20 seconds is loading of unnecessary profiles (yes, mine profiles are big - hundreds of controls, almost thousand checks).

**Comment by voroniys** [comment]
Just found yet another issue with the blind pre-loading of all dependencies - these profiles could contain some ruby code that will just fail if executed on wrong system, just because some softweare is not installed.
This is even more serious issue, that previous performance one, since the whole execution has failed without executing a single check.
